src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -0,0 +1,536 @@
+    internal static class SseIntrinsics
+    {
+        private static unsafe Vector128<float> Load1(float* src, int* idx)
+        {
+            return Sse.SetScalarVector128(src[idx[0]]);
+        }
+
+        private static unsafe Vector128<float> Load4(float* src, int* idx)
+        {
+            return Sse.SetVector128(src[idx[3]], src[idx[2]], src[idx[1]], src[idx[0]]);
+        }
+
+        private static Vector128<float> Rotate(Vector128<float> x)
+        {
+            return Sse.Shuffle(x, x, 0x39);
+        }
+
+        private static Vector128<float> RotateReverse(Vector128<float> x)
+        {
+            return Sse.Shuffle(x, x, 0x93);
+        }
+
+        // Warning: this operation changes the value of x => do not reuse x
+        private static unsafe void Store4(Vector128<float> x, float* dst, int* idx)
+        {
+            Sse.StoreScalar(dst + idx[0], x);
+            for (int i = 1; i <= 3; i++)
+            {
+                x = Rotate(x);
+                Sse.StoreScalar(dst + idx[i], x);
+            }
+        }
+
+        internal static void ScaleU(float scale, float[] dst, int count)
+        {
+            Vector128<float> scaleVector = Sse.SetAllVector128(scale);
+
+            unsafe
+            {
+                fixed (float* pdst = dst)
+                {
+                    float* pDstCurrent = pdst;
+                    float* pEnd = pdst + count;
+
+                    while (pDstCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> dstVector = Sse.LoadVector128(pDstCurrent);
+
+                        dstVector = Sse.Multiply(scaleVector, dstVector);
+                        Sse.Store(pDstCurrent, dstVector);
+
+                        pDstCurrent += 4;
+                    }
+
+                    while (pDstCurrent < pEnd)
+                    {
+                        Vector128<float> dstVector = Sse.LoadScalarVector128(pDstCurrent);
+
+                        dstVector = Sse.MultiplyScalar(scaleVector, dstVector);
+                        Sse.StoreScalar(pDstCurrent, dstVector);
+
+                        pDstCurrent++;
+                    }
+                }
+            }
+        }
+
+        internal static void AddScaleU(float scale, float[] src, float[] dst, int count)
+        {
+            Vector128<float> scaleVector = Sse.SetAllVector128(scale);
+
+            unsafe
+            {
+                fixed (float* psrc = src)
+                fixed (float* pdst = dst)
+                {
+                    float* pSrcCurrent = psrc;
+                    float* pDstCurrent = pdst;
+                    float* pEnd = pdst + count;
+
+                    while (pDstCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadVector128(pSrcCurrent);
+                        Vector128<float> dstVector = Sse.LoadVector128(pDstCurrent);
+
+                        srcVector = Sse.Multiply(srcVector, scaleVector);
```

Comment:
On newer hardware, this can become a `FusedMultiplyAdd`, provided a different (but more precise) result is acceptable for some inputs.

Reply:
It would be part of any AVX related code-work.